### PR TITLE
Support snippet text edit

### PR DESCRIPTION
**Pull request by matklad**
bors r+
🤖

### CODE REVIEW

File: crates/rust-analyzer/src/main_loop/handlers.rs [hunk 1]
```diff
@@ -696,14 +694,21 @@ pub fn handle_formatting(
 pub fn handle_code_action(
     world: WorldSnapshot,
     params: lsp_types::CodeActionParams,
-) -> Result<Option<CodeActionResponse>> {
+) -> Result<Option<Vec<lsp_ext::CodeAction>>> {
     let _p = profile("handle_code_action");
+    // We intentionally don't support command-based actions, as those either
+    // requires custom client-code anyway, or requires server-initiated edits.
+    // Server initiated edits break causality, so we avoid those as well.
+    if !world.config.client_caps.code_action_literals {
```
Comment (matklad) [inline crates/rust-analyzer/src/main_loop/handlers.rs:702]: cc @kjeremy 

I think, given that pure commands require us to implement client-specific logic anyway, I figured that we might as well just not support them? This is related to the caps PR

Comment (kjeremy) [reply crates/rust-analyzer/src/main_loop/handlers.rs:702]: I think that's only true until we support `workspace/executeCommand`.

Comment (matklad) [reply crates/rust-analyzer/src/main_loop/handlers.rs:702]: https://github.com/microsoft/language-server-protocol/issues/724 is the upstream issue. 

>I think that's only true until we support `workspace/executeCommand`.

I personally think that we should not support this :-) The only command that the server can execute is sending workspace edit to the client, but that's strictly worse than just sending it with edit. 

Ie, I think it is confirming to the protocol and a good tradeoff to just not support clients that can do commands with edits.

File: docs/dev/lsp-extensions.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+```typescript
+export interface TextDocumentEdit {
+	textDocument: VersionedTextDocumentIdentifier;
+	edits: (TextEdit | SnippetTextEdit)[];
+}
+```
+
+When applying such code action, the editor should insert snippet, with tab stops and placeholder.
+At the moment, rust-analyzer guarantees that only a single edit will have `InsertTextFormat.Snippet`.
```
[the remark this answers is not in the document]
Comment (matklad) [reply docs/dev/lsp-extensions.md:34]: https://github.com/microsoft/language-server-protocol/issues/724 is the upstream issue.

**Comment by fannheyward** [comment]
coc-rust-analyzer has added snippet text edit support. https://github.com/fannheyward/coc-rust-analyzer/pull/252, anyone can give this a test?

File: docs/dev/lsp-extensions.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (matklad) [reply docs/dev/lsp-extensions.md:34]: created https://github.com/rust-analyzer/rust-analyzer/issues/4604 as a more structured way to send pings :D

**Pull request merged**